src/storage/v2/inmemory/storage.cpp

```diff
@@ -815,19 +822,31 @@ utils::BasicResult<StorageManipulationError, void> InMemoryStorage::InMemoryAcce
 
       if (storage_->config_.salient.items.enable_label_index_auto_creation) {
         storage_->labels_to_auto_index_.WithLock([&](auto &label_indices) {
-          for (const auto label : label_indices) {
-            CreateIndex(label, false);
+          for (auto &label : label_indices) {
+            --label.second;
+            // If there are multiple transactions that would like to create an
+            // auto-created index on a specific label, we only build the index
+            // when the last one commits.
+            if (label.second == 0) {
+              CreateIndex(label.first, false);
```

Comment:
Why do we create only during last commit instead of the first commit?

Reply:
It is possible that 2 or more concurrent transactions are trying to insert a vertex, attempting to auto-create the same index. There are 3 steps in auto-creating a vertex
 1. Update the metadata data-structure that holds the indices we would like to create during commit.
 2. Call that function that would update the indexes with the current vertex.(This will not fire in either case since the index is not yet existing meaning that at this point there is only a set of vertices that may or may not be commited, nothing in the indices).
 3. Create the index by during commit.
 
 It is possible that `TXN1` is at step 3, while `TXN2` just finished with step 2, but not yet at step 3. If that is the case the index will only contain the vertex from `TXN1`, since `TXN2` just missed the opportunity the update the index, since it does not exist yet.
 
 If we create the index on the last active transaction that is introducing that specific label or edge-type, the index will be built correctly without missing any entries.

Reply:
The problem I see here is following.
0. Auto label index creation is enabled.
1. Assume a streaming app which inserts data to Memgraph non-stop with the same label which previously hasn't been indexed.
2. In parallel somebody runs read transactions by this new label. This will always do ScanAll instead of ScanByLabelIndex.
3. Label index will not be created because there are infinite parallel write transactions.

Currently I don't see a proper solution without losing data entries. Maybe we should update indices during commit, but this will affect reads during the same transaction when in ScanByLabelIndex we will not see vertices added in the same txn.

So I would say let's keep it now as it is, because this will give correct results. And think later can we improve there or not.